lib/accesslists/hierarchy.go

```diff
@@ -440,54 +441,235 @@ func UserMeetsRequirements(identity types.User, requires accesslist.Requires) bo
+func (s *Hierarchy) membershipAncestors(ctx context.Context, list *accesslist.AccessList, user types.User) ([]*accesslist.AccessList, error) {
+	return getAncestors(ctx, list, RelationshipKindMember, s.AccessListsService, withUserRequirementsCheck(user, s.Clock))
+}
+
+func (s *Hierarchy) expandOwnerOf(ctx context.Context, ancestors []*accesslist.AccessList, user types.User) ([]*accesslist.AccessList, error) {
+	var out []*accesslist.AccessList
+	for _, v := range ancestors {
+		for _, name := range v.Status.OwnerOf {
+			lst, err := s.AccessListsService.GetAccessList(ctx, name)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			if UserMeetsRequirements(user, lst.GetOwnershipRequires()) {
+				out = append(out, lst)
+			}
+		}
+	}
+	return out, nil
+}
+
+func (s *Hierarchy) validDirectOwner(user types.User, acl *accesslist.AccessList) bool {
+	if !UserMeetsRequirements(user, acl.GetOwnershipRequires()) {
+		return false
+	}
+	for _, v := range acl.Spec.Owners {
+		if v.Name == user.GetName() {
+			return true
+		}
+	}
+	return false
+}
+
+func getAncestors(ctx context.Context, accessList *accesslist.AccessList, kind RelationshipKind, g AccessListAndMembersGetter, opts ...ancestorOption) ([]*accesslist.AccessList, trace.Error) {
 	ancestorsMap := make(map[string]*accesslist.AccessList)
-	if err := collectAncestors(ctx, accessList, kind, g, make(map[string]struct{}), ancestorsMap); err != nil {
+	if err := collectAncestors(ctx, accessList, kind, g, make(map[string]struct{}), ancestorsMap, opts...); err != nil {
 		return nil, trace.Wrap(err)
 	}
 	ancestors := slices.Collect(maps.Values(ancestorsMap))
 	return ancestors, nil
 }
 
-func collectAncestors(ctx context.Context, accessList *accesslist.AccessList, kind RelationshipKind, g AccessListAndMembersGetter, visited map[string]struct{}, ancestors map[string]*accesslist.AccessList) error {
+// GetAncestorsFor calculates and returns the set of Ancestor ACLs depending on
+// the supplied relationship criteria. Order of the ancestor list is undefined.
+func GetAncestorsFor(ctx context.Context, accessList *accesslist.AccessList, kind RelationshipKind, g AccessListAndMembersGetter) ([]*accesslist.AccessList, error) {
+	return getAncestors(ctx, accessList, kind, g)
+}
+
+func collectAncestors(ctx context.Context, accessList *accesslist.AccessList, kind RelationshipKind, g AccessListAndMembersGetter, visited map[string]struct{}, ancestors map[string]*accesslist.AccessList, opts ...ancestorOption) error {
+	options := &ancestorOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+	if err := options.validate(); err != nil {
+		return trace.Wrap(err)
+	}
 	if _, ok := visited[accessList.GetName()]; ok {
 		return nil
 	}
 	visited[accessList.GetName()] = struct{}{}
 
-	switch kind {
-	case RelationshipKindOwner:
+	isDirectMembershipExpired := func(acl, member string) (bool, error) {
+		if !options.validateUserRequirement {
```

Comment:
Curious on which scenario do we want to skip user requirement validation?

Reply:
The validationUserRequirement option makes it clearer that user requirements should be explicitly validated.
The interface sets `validateUserRequirement` whenever a user object is provided.

I did consider using a simple `if user != nil { ... }`, but that approach feels less clear and would require an explanatory comment. In contrast, having options.validateUserRequirement makes the intent self-explanatory.